## 기술통계 연습문제 (2)

### #01. 작업준비

#### 필요한 패키지 설치

In [2]:
import sys #글꼴처리를 위한 패키지
import numpy as np
import seaborn as sb
from pandas import DataFrame, read_excel
from matplotlib import pyplot as plt

#### 데이터 불러오기

In [3]:
df = read_excel("https://data.hossam.kr/D02/wage.xlsx")
df

year  age            maritl      race        education  \
0     2006   18  1. Never Married  1. White     1. < HS Grad   
1     2004   24  1. Never Married  1. White  4. College Grad   
2     2003   45        2. Married  1. White  3. Some College   
3     2003   43        2. Married  3. Asian  4. College Grad   
4     2005   50       4. Divorced  1. White       2. HS Grad   
...    ...  ...               ...       ...              ...   
2995  2008   44        2. Married  1. White  3. Some College   
2996  2007   30        2. Married  1. White       2. HS Grad   
2997  2005   27        2. Married  2. Black     1. < HS Grad   
2998  2005   27  1. Never Married  1. White  3. Some College   
2999  2009   55      5. Separated  1. White       2. HS Grad   

                  region        jobclass          health health_ins   logwage  \
0     2. Middle Atlantic   1. Industrial       1. <=Good      2. No  4.318063   
1     2. Middle Atlantic  2. Information  2. >=Very Good      2. No  4.255273   
2     2. Middle Atlantic   1. Industrial       1. <=Good     1. Yes  4.875061   
3     2. Middle Atlantic  2. Information  2. >=Very Good     1. Yes  5.041393   
4     2. Middle Atlantic  2. Information       1. <=Good     1. Yes  4.318063   
...                  ...             ...             ...        ...       ...   
2995  2. Middle Atlantic   1. Industrial  2. >=Very Good     1. Yes  5.041393   
2996  2. Middle Atlantic   1. Industrial  2. >=Very Good      2. No  4.602060   
2997  2. Middle Atlantic   1. Industrial       1. <=Good      2. No  4.193125   
2998  2. Middle Atlantic   1. Industrial  2. >=Very Good     1. Yes  4.477121   
2999  2. Middle Atlantic   1. Industrial       1. <=Good     1. Yes  4.505150   

            wage  
0      75.043154  
1      70.476020  
2     130.982177  
3     154.685293  
4      75.043154  
...          ...  
2995  154.685293  
2996   99.689464  
2997   66.229408  
2998   87.981033  
2999   90.481913  

[3000 rows x 11 columns]

In [4]:
df.dtypes

year            int64
age             int64
maritl         object
race           object
education      object
region         object
jobclass       object
health         object
health_ins     object
logwage       float64
wage          float64
dtype: object

#### #02. 데이터를 로드하여 명목형 변수를 1, 2 등으로 레이블링 하시오. 값의 종류는 데이터프레임으로부터 조회하여 확인하시오.

In [34]:
# 명목형 변수 : 순서, 크기 X 

# maritl
df['maritl'] = df['maritl'].str.extract(r'(\d+)')

# # race
df['race'] = df['race'].str.extract(r'(\d+)')

# # education
df['education'] = df['education'].str.extract(r'(\d+)')

# region
df['region'] = df['region'].str.extract(r'(\d+)')

# jobclass
df['jobclass'] = df['jobclass'].str.extract(r'(\d+)')

# health
df['health'] = df['health'].str.extract(r'(\d+)')


# health_ins
df['health_ins'] = df['health_ins'].str.extract(r'(\d+)')

df

year  age maritl race education region jobclass health health_ins  \
0     2006   18      1    1         1      2        1      1          2   
1     2004   24      1    1         4      2        2      2          2   
2     2003   45      2    1         3      2        1      1          1   
3     2003   43      2    3         4      2        2      2          1   
4     2005   50      4    1         2      2        2      1          1   
...    ...  ...    ...  ...       ...    ...      ...    ...        ...   
2995  2008   44      2    1         3      2        1      2          1   
2996  2007   30      2    1         2      2        1      2          2   
2997  2005   27      2    2         1      2        1      1          2   
2998  2005   27      1    1         3      2        1      2          1   
2999  2009   55      5    1         2      2        1      1          1   

       logwage        wage  
0     4.318063   75.043154  
1     4.255273   70.476020  
2     4.875061  130.982177  
3     5.041393  154.685293  
4     4.318063   75.043154  
...        ...         ...  
2995  5.041393  154.685293  
2996  4.602060   99.689464  
2997  4.193125   66.229408  
2998  4.477121   87.981033  
2999  4.505150   90.481913  

[3000 rows x 11 columns]

##### 명목형 변수 레이블링
<maritl>
- 방법1
vc = df['maritl'].values_counts()
vlist = list(vc.index)
vlist

conditions = [
    (df['maritl']==vlist[0]),
    (df['maritl']==vlist[1]),
    (df['maritl']==vlist[2]),
    (df['maritl']==vlist[3]),
    (df['maritl']==vlist[4]),
]

value = [2,1,4,5,3]

df['maritl'] = np.select(conditions, value)

- 방법2
df.loc[df['maritl'] == 1. Never Married, 'maritl'] = 1
df

----------------------------------------------------
<반복문> : 명목형변수를 찾아서 카테고리형으로 바꾸는 반복문

df.dtypes.index => 컬럼 이름만
df.dtypes.values => 데이터의 값만

cdf = df.copy()
ilist = list(df.dtypes.index)
vlist = list(df.dtypes.values)

for i, v in enumertate(vlist):
    if v == 'object':
        field_name = ilist[i]
        <!-- print(filedname) -->

        vc = df[field_name].value_counts()
        <!-- print(vs) -->
        또는
        vc = cdf[field_name].value_counts().sort_index() (=count개수가 많은 순으로 출력)

        for ii, vv in enumerate(list(vs.index)):
            p = vv.find(".")
            vnum = int(vv[:p]) (자료의 숫자값만 도출하는 방법)
            
            또는
            vnum = ii + 1 (인덱스번호에 +1 시켜서 정렬시키는 방법)

            cdf.loc[df['field_name'] == vv, field_name] = vnum

        cdf[field_name] = df[field_name].astype('category')
cdf

#### #03. 레이블링 된 명목형 변수를 category 타입으로 변경하시오.

In [6]:
df['maritl'] = df['maritl'].astype('category')
df['race'] = df['race'].astype('category')
df['education'] = df['education'].astype('category')
df['region'] = df['region'].astype('category')
df['jobclass'] = df['jobclass'].astype('category')
df['health'] = df['health'].astype('category')
df['health_ins'] = df['health_ins'].astype('category')
df.dtypes


year             int64
age              int64
maritl        category
race          category
education     category
region        category
jobclass      category
health        category
health_ins    category
logwage        float64
wage           float64
dtype: object

#### #04. 수치형 변수에 대한 요약 통계를 확인하고 설명하시오 (상자그림 제외)

In [20]:
# 수치형 변수 
desc = df[['age','logwage','wage']].describe()
desc

age      logwage         wage
count  3000.000000  3000.000000  3000.000000
mean     42.414667     4.653905   111.703608
std      11.542406     0.351753    41.728595
min      18.000000     3.000000    20.085537
25%      33.750000     4.447158    85.383940
50%      42.000000     4.653213   104.921507
75%      51.000000     4.857332   128.680488
max      80.000000     5.763128   318.342430

df.describe() 하면 끝... #명목형은 알아서 빠지게 해준다.

In [8]:
# 설명
# 조사인원은 3000명이다. 
# age : 하위 25%의 나이는 33.75세이며, 중앙값은 42세, 하위 75%는 51세이다.
# logwage : 임금로그값의 표준편차는 0.351753이고, 평균은 4.653905이다. 
# wage : 임금의 평균은 111.703608이며, 표준편차는 41.728595이다.
# 연령에 대한 표준편차는 11.5이고, 임금의 표준편차는 41.73이다.
# 평균은 111.7인 반면 표준편차는 41.73이므로, 표준편차가 크다. (= 퍼짐의 정도가 크다.)

결과분석의 예.
표준편차가 크다 = 데이터의 범위가 크다. (퍼짐의 정도가 크다)

#### #05. 명목형 변수에 대한 기술 통계를 수행하고 설명하시오.

* value_counts() 자체가 명목형 변수 기술통계 (명목형변수는 빈도수로 측정한다.)

cdf = df.copy()
cdf2 = cdf.drop(['year','age','lowage','wage'], axis = 1) #명목형변수만 나옴

for v in cdf2.columns:
    print(cdf2[v].value_counts()) #기술통계는 value_counts()자체
    print(#--------------------#)

In [9]:
desc2 = df.groupby(['maritl']).describe()
desc2

year                                                                 \
         count         mean       std     min     25%     50%     75%     max   
maritl                                                                          
1        648.0  2005.856481  2.054005  2003.0  2004.0  2006.0  2008.0  2009.0   
2       2074.0  2005.779171  2.019545  2003.0  2004.0  2006.0  2008.0  2009.0   
3         19.0  2005.157895  1.675416  2003.0  2004.0  2005.0  2006.5  2009.0   
4        204.0  2005.715686  2.019010  2003.0  2004.0  2005.5  2007.0  2009.0   
5         55.0  2005.963636  2.090221  2003.0  2004.5  2006.0  2008.0  2009.0   

           age             ...   logwage              wage              \
         count       mean  ...       75%       max   count        mean   
maritl                     ...                                           
1        648.0  32.583333  ...  4.681241  5.750441   648.0   92.734649   
2       2074.0  44.861620  ...  4.903090  5.763128  2074.0  118.860261   
3         19.0  50.789474  ...  4.770790  4.954243    19.0   99.538655   
4        204.0  47.607843  ...  4.772677  5.641006   204.0  103.159261   
5         55.0  43.818182  ...  4.720140  5.626900    55.0  101.215787   

                                                                             
              std        min        25%         50%         75%         max  
maritl                                                                       
1       32.920068  23.274704  73.775743   87.981033  107.903924  314.329336  
2       43.120554  20.085537  91.699226  112.648896  134.705375  318.342430  
3       23.741055  54.598150  86.682487  104.921507  118.015588  141.775172  
4       33.800978  22.962401  83.704378   99.689464  118.235905  281.745971  
5       33.663381  61.187526  81.283253   95.230713  112.184873  277.799480  

[5 rows x 32 columns]

> 분석결과 <br>
결혼한 사람들이 많으며, 결혼한 사람들의 임금 평균이 가장 높았다.

In [10]:
desc3 = df.groupby(['race']).describe()
desc3

year                                                                 \
       count         mean       std     min     25%     50%     75%     max   
race                                                                          
1     2480.0  2005.757258  2.022451  2003.0  2004.0  2006.0  2008.0  2009.0   
2      293.0  2005.924915  1.951771  2003.0  2004.0  2006.0  2008.0  2009.0   
3      190.0  2006.026316  2.163757  2003.0  2004.0  2006.0  2008.0  2009.0   
4       37.0  2005.783784  2.083468  2003.0  2004.0  2005.0  2008.0  2009.0   

         age             ...   logwage              wage              \
       count       mean  ...       75%       max   count        mean   
race                     ...                                           
1     2480.0  42.394758  ...  4.867864  5.750441  2480.0  112.563666   
2      293.0  43.563140  ...  4.763428  5.763128   293.0  101.601177   
3      190.0  41.821053  ...  4.975547  5.701323   190.0  120.288288   
4       37.0  37.702703  ...  4.653213  5.133021    37.0   89.973326   

                                                                           
            std        min        25%         50%         75%         max  
race                                                                       
1     41.733830  20.085537  86.695155  106.348256  130.042934  314.329336  
2     37.162486  23.274704  81.283253   94.072715  117.146817  318.342430  
3     46.422514  34.607418  87.981033  115.375039  144.828129  299.262977  
4     29.153530  38.223583  71.319639   81.283253  104.921507  169.528538  

[4 rows x 32 columns]

> 분석결과 <br>
백인이 가장 많으며, 백인의 평균임금보다 아시아인의 평균임금이 더 높았다.

In [11]:
desc4 = df.groupby(['education']).describe()
desc4

year                                                         \
           count         mean       std     min     25%     50%     75%   
education                                                                 
1          268.0  2005.764925  2.104205  2003.0  2004.0  2006.0  2008.0   
2          971.0  2005.806385  2.022451  2003.0  2004.0  2006.0  2008.0   
3          650.0  2005.695385  1.955470  2003.0  2004.0  2006.0  2007.0   
4          685.0  2005.795620  2.050313  2003.0  2004.0  2006.0  2008.0   
5          426.0  2005.910798  2.053754  2003.0  2004.0  2006.0  2008.0   

                     age             ...   logwage             wage  \
              max  count       mean  ...       75%       max  count   
education                            ...                              
1          2009.0  268.0  41.794776  ...  4.579784  5.025306  268.0   
2          2009.0  971.0  42.217302  ...  4.698970  5.763128  971.0   
3          2009.0  650.0  40.887692  ...  4.798995  5.750441  650.0   
4          2009.0  685.0  42.773723  ...  4.963788  5.641006  685.0   
5          2009.0  426.0  45.007042  ...  5.144566  5.763128  426.0   

                                                                     \
                 mean        std        min         25%         50%   
education                                                             
1           84.104415  21.578051  20.934378   70.261774   81.283253   
2           95.783351  28.567560  23.274704   77.946382   94.072715   
3          107.755566  32.474725  20.085537   89.242885  104.921507   
4          124.427908  41.189071  32.366413   99.689464  118.884359   
5          150.917776  53.904206  38.605915  117.146817  141.775172   

                                   
                  75%         max  
education                          
1           97.493294  152.216807  
2          109.833986  318.342430  
3          121.388422  314.329336  
4          143.134941  281.745971  
5          171.497631  318.342430  

[5 rows x 32 columns]

> 분석결과 <br>
HS Grad가 가장 많으며, 평균임금이 가장 높았다.

In [12]:
desc5 = df.groupby(['region']).describe()
desc5

year                                                              \
         count      mean       std     min     25%     50%     75%     max   
region                                                                       
2       3000.0  2005.791  2.026167  2003.0  2004.0  2006.0  2008.0  2009.0   

           age             ...   logwage              wage              \
         count       mean  ...       75%       max   count        mean   
region                     ...                                           
2       3000.0  42.414667  ...  4.857332  5.763128  3000.0  111.703608   

                                                                           
              std        min       25%         50%         75%        max  
region                                                                     
2       41.728595  20.085537  85.38394  104.921507  128.680488  318.34243  

[1 rows x 32 columns]

> 분석결과 <br>
Middle Atlantic 지역의 평균임금은 111.703608이며 하위 25%는 85.38394, 중앙값은 104.921507, 하위 75%는 128.680488 이다.

In [13]:
desc6 = df.groupby(['jobclass']).describe()
desc6

year                                                         \
           count         mean       std     min     25%     50%     75%   
jobclass                                                                  
1         1544.0  2005.803109  2.048053  2003.0  2004.0  2006.0  2008.0   
2         1456.0  2005.778159  2.003321  2003.0  2004.0  2006.0  2007.0   

                     age             ...   logwage              wage  \
             max   count       mean  ...       75%       max   count   
jobclass                             ...                               
1         2009.0  1544.0  41.398316  ...  4.778151  5.690330  1544.0   
2         2009.0  1456.0  43.492445  ...  4.924279  5.763128  1456.0   

                                                                               \
                mean        std        min        25%         50%         75%   
jobclass                                                                        
1         103.321129  35.110674  22.962401  81.283253   99.689464  118.884359   
2         120.592721  46.131439  20.085537  91.699226  112.648896  137.590143   

                     
                max  
jobclass             
1         295.99125  
2         318.34243  

[2 rows x 32 columns]

> 분석결과 <br>
Industrial 종사자들의 평균임금은 103.321129이며, Information종사자들의 평균임금은 120.592721이다.
> Information에 종사하는 사람들이 Industrial에 종사하는 사람들보다 임금이 높다.

In [14]:
desc7 = df.groupby(['health']).describe()
desc7

year                                                                 \
         count         mean       std     min     25%     50%     75%     max   
health                                                                          
1        858.0  2005.797203  2.009520  2003.0  2004.0  2006.0  2007.0  2009.0   
2       2142.0  2005.788515  2.033259  2003.0  2004.0  2006.0  2008.0  2009.0   

           age             ...   logwage              wage              \
         count       mean  ...       75%       max   count        mean   
health                     ...                                           
1        858.0  44.947552  ...  4.778151  5.742793   858.0  101.661285   
2       2142.0  41.400093  ...  4.903090  5.763128  2142.0  115.726164   

                                                                             
              std        min        25%         50%         75%         max  
health                                                                       
1       35.184998  20.085537  79.854900   98.599344  118.884359  311.934569  
2       43.438963  20.934378  87.981033  109.833986  134.705375  318.342430  

[2 rows x 32 columns]

> 분석결과 <br>
건강상태가 'good'인 경우의 평균 임금은 101.661285이고, 'very good'인 경우 평균임금은 115.726164이다. 
> 건강상태가 좋은 사람의 임금이 더 높았다. 

#### #06. 결혼 여부에 따른 임금 수준을 비교하고자 한다. 결혼 여부에 따라 서브플롯을 구성하여 임금 수준을 히스토그램으로 시각화 하고 설명하시오.

In [15]:
# 한글 폰트
plt.rcParams['font.family'] = 'Malgun Gothic'
plt.rcParams['font.size'] = 10
plt.rcParams['figure.figsize'] = (12,8)
plt.rcParams['axes.unicode_minus'] = False

나는.. 데이터프레임을 분리할 생각함..

선생님 풀이
#  ax1 데이터 추출 & 히스토그램 예시...
mdf = cdf.query("maritl == 1")
hist, bins = np.histogram(mdf['wage'], bins=10)
sb.histplot(data=mdf, x='wage', bins=10, ax=ax[0])
ax[0].set_title('maritl=1')
ax[0].set_xticks(bins)
ax[0].set_xticklabels(bins, rotation=30)

# 예시를 바탕으로 ~
# 서브플롯 영역 설정
fig, ax = plt.subplots (len(v.index), 1, figsize = (15,30))

# 데이터 추출 및 히스토그램 반복문
vcount = cdf['maritl'].value_counts().sort_index()
vcount.index

for i, v in enumerate(vcount.index):
    mdf = cdf.query("maritl == %v")
    hist, bins = np.histogram(mdf['wage'], bins=10)
    sb.histplot(data=mdf, x='wage', bins=10, ax=ax[i])
    ax[i].set_title('maritl==%d' %v)
    ax[i].set_xticks(bins)
    ax[i].set_xticklabels(bins, rotation=30)


plt.show()
plt.close()


maritl -> 인덱스가 5개 이므로, 5개의 서브플롯을 만드는 것임.

#### #07. 교육 수준에 따른 임금에 대한 히스토그램을 시각화 하고 설명하시오. 교육수준별로 그래프를 나누어 서브플롯으로 제시해야 합니다.

#### #08. 직군별 건강상태를 확인하고자 한다. 적절한 형태로 데이터를 재배치하고 설명하시오.

* 풀이 1
필드가 하나 더 있다면 -> pivot_table활용
hdf = cdf.filter['jobclass','health'] + age 필드
pivot_table(hdf, index='jobclass', columns='health', aggfunc='count')
근데 약간 어거지스러움...

필요한 필드만 가져오려면 -> crosstab 활용
hdf2 = cdf.filter['jobclass','health'] 
crosstab(index=hdf2['jobclass'], columns=hdf2['health'],
    rownames=['직군'], colnames=['건강상태'], margins=True)


#### #09.교육 수준을 인종 비율에 따라 설명하고자 한다. 적절한 시각화 자료를 제시하고 설명하시오.

교육수준 = data, 인종비율에 따라 = hue